xc/xenops_server_xen.ml

```diff
@@ -2551,27 +2561,31 @@ module VIF = struct
          try
            (* If the device is gone then this is ok *)
            let device = device_by_id xc xs vm Device_common.Vif Oldest (id_of vif) in
-           let destroy device =
-             (* NB different from the VBD case to make the test pass for now *)
-             Xenops_task.with_subtask task (Printf.sprintf "Vif.hard_shutdown %s" (id_of vif))
-               (fun () -> (if force then Device.hard_shutdown else Device.clean_shutdown) task ~xs device);
-             Xenops_task.with_subtask task (Printf.sprintf "Vif.release %s" (id_of vif))
-               (fun () -> Device.Vif.release task ~xc ~xs device) in
-           destroy device;
-
-           Opt.iter (fun vm_t ->
-               (* If we have a qemu frontend, detach this too. *)
-               if List.mem_assoc vif.Vif.id vm_t.VmExtra.non_persistent.VmExtra.qemu_vifs then begin
-                 match (List.assoc vif.Vif.id vm_t.VmExtra.non_persistent.VmExtra.qemu_vifs) with
-                 | _, Device device ->
-                   destroy device;
-                   let non_persistent = { vm_t.VmExtra.non_persistent with
-                                          VmExtra.qemu_vifs = List.remove_assoc vif.Vif.id vm_t.VmExtra.non_persistent.VmExtra.qemu_vifs } in
-                   DB.write vm { vm_t with VmExtra.non_persistent = non_persistent }
-                 | _, _ -> ()
-               end;
-             ) vm_t;
-
+           match vif.backend with
+           | Network.Local _ | Network.Remote _ ->
+             (* If the device is gone then this is ok *)
+             let destroy device =
+               (* NB different from the VBD case to make the test pass for now *)
+               Xenops_task.with_subtask task (Printf.sprintf "Vif.hard_shutdown %s" (id_of vif))           
+                 (fun () -> (if force then Device.hard_shutdown else Device.clean_shutdown) task ~xs device);
+               Xenops_task.with_subtask task (Printf.sprintf "Vif.release %s" (id_of vif))
+                 (fun () -> Device.Vif.release task ~xc ~xs device) in
+             destroy device;
+             Opt.iter (fun vm_t ->
+                 (* If we have a qemu frontend, detach this too. *)
+                 if List.mem_assoc vif.Vif.id vm_t.VmExtra.non_persistent.VmExtra.qemu_vifs then begin
+                   (match (List.assoc vif.Vif.id vm_t.VmExtra.non_persistent.VmExtra.qemu_vifs) with
```

Comment:
You don't need the additional set of parentheses around `match ... with ...` when they are wrapped by `begin ... end`

Reply:
Fixed. Thanks.